Refactoring of BeamSpotProblemMonitor

Dr15Jones opened the pull request:
-Converted to a DQMEDAnalyzer
-removed unnecessary use of std::map
-removed unnecessary code
-Conformed to CMS naming convention
-Converted to modern C++ conventions

Dr15Jones committed:
Refactoring of BeamSpotProblemMonitor

-Converted to a DQMEDAnalyzer
-removed unnecessary use of std::map
-removed unnecessary code
-Conformed to CMS naming convention
-Converted to modern C++ conventions

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
@knoepfel FYI

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24336/6079

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

DQM/BeamMonitor


@kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please review it and eventually sign? Thanks.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
This can now be a DQMEDAnalyzer since the base class is now a `one` module. This allows BeamSpotProblemMonitor to see all the events, which is what it wants.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
An issue related to this code is #24337

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -1,168 +1,130 @@
+  //NOTE: This in principal should be a Lumi only histogram since it gets reset at every
+  // beginLuminosityBlock call. However, it is also filled at that time and the DQMStore
+  // clears all lumi histograms at postGlobalBeginLuminosityBlock!
+  beamSpotError_ = iB.book1D("BeamSpotError","ERROR: Beamspot missing from scalars",20,0.5,20.5);
```
As the comment says, this really should have `setLumiFlag()` called on it, but that would not work because of the exact timing per lumi flags get reset.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
+  bool allTkOn = true;
+  //BUG??? DcsStatus::TECm is ignored
+  for (int i=0; i<5; i++) {
+    if (!dcsTk[i]) {
```
This appears to be a bug from the original version of the code. If you want it fixed, I'd change it to
```
for(auto e: dcsTk) {
   if(!e) {
      allTkOn = false;
      break;
   }
}
```
Which will work even if the size if `dcsTk` is ever changed.

andrius-k replied:
@sushilchauhan FYI. You seem to be the author of this code so maybe you would be qualified to answer this question.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
+  // beam spot scalar object
+  BeamSpotOnline spotOnline;
+
+  bool fallBackToDB=false;
```
I find it odd that if `handleScaler->empty()` is `true` then this code does not fall back to the DB as it will do if it thinks the fit result is not valid. Is that the actual intended behavior?

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
-                      }
 
+  //Get quality report
+  const QReport * beamSpotQReport = beamSpotError_->getQReport("BeamSpotOnlineTest");
```
Does this actually do anything useful? I was concerned this code is intended to cause a side-effect which is why I did not remove it.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
 
-     const QReport * BeamSpotQReport = myQReport->getQReport("BeamSpotOnlineTest");  
+   //Reset it here for next lumi
+  beamSpotError_->Reset();
```
This would go away if I could call `setLumiFlag()` on it.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
-}
-
-//--------------------------------------------------------
-void BeamSpotProblemMonitor::endJob(const LuminosityBlock& lumiSeg,
```
This `endJob` function was never called by the framework because it had the wrong arguments.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.cc:
```diff
@@ -189,182 +150,156 @@ void BeamSpotProblemMonitor::analyze(const Event& iEvent,
-     BeamSpotError->Reset();
+  if(debug_)edm::LogInfo("BeamSpotProblemMonitor") << "endRun:: Clearing all the Maps "<<endl;
+  //Reset it end of job
+  beamSpotError_->Reset();
```
I'm not sure this `Reset` has any practical value since it will just get reset again at the first LuminosityBlock of the next Run.

Dr15Jones commented on DQM/BeamMonitor/plugins/BeamSpotProblemMonitor.h:
```diff
@@ -11,86 +11,71 @@
     // EndRun
     void endRun(const edm::Run& r, const edm::EventSetup& c) override;
-    // Endjob
-    void endJob(const edm::LuminosityBlock& lumiSeg, const edm::EventSetup& c);
```
Not being able to add `override` to this function should have been a clue that there was a problem.

cmsbuild commented:
+1
Tested at: 16369555b58068545b9051d4b13a49c3e57f39fe
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24336/29911/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24336/29911/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2983602
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2983410
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

andrius-k commented:
@dinardo Maybe you want to answer some of the questions here?

dinardo commented:
Dear Andrius,
I'm not sure why you contacted me.
As far as I understood the changes here refer to the files:
- BeamSpotProblemMonitor.cc
- BeamSpotProblemMonitor.h
- BeamSpotProblemMonitor_cff.py
I'm not involved in the development of this package.

I'm involved in the development of the package:
BeamPixel application, which is composed of the following files:
https://github.com/dinardo/cmssw/tree/master/DQM/BeamMonitor/plugins
- Vx3DHLTAnalyzer.cc
- Vx3DHLTAnalyzer.h
and
https://github.com/dinardo/cmssw/blob/master/DQM/Integration/python/clients/beampixel_dqm_sourceclient-live_cfg.py
(together with its rendering plugin)

Please let me know if I misunderstood your question.

Kind regards.
- Mauro.

sushilchauhan commented:
Hi Andrius,

On 20/08/18 9:24 PM, Andrius wrote:
> andrius-k commented on this pull request.
>
>
>
>> +  iEvent.getByToken(dcsStatus_, dcsStatus);
> +  std::array<bool,6> dcsTk;
> +  for (auto &e: dcsTk) { e=true; }
> +  for (auto const& status: *dcsStatus ) {
> +    if (!status.ready(DcsStatus::BPIX))   dcsTk[0]=false;
> +    if (!status.ready(DcsStatus::FPIX))   dcsTk[1]=false;
> +    if (!status.ready(DcsStatus::TIBTID)) dcsTk[2]=false;
> +    if (!status.ready(DcsStatus::TOB))    dcsTk[3]=false;
> +    if (!status.ready(DcsStatus::TECp))   dcsTk[4]=false;
> +    if (!status.ready(DcsStatus::TECm))   dcsTk[5]=false;
> +  }
> +
> +  bool allTkOn = true;
> +  //BUG??? DcsStatus::TECm is ignored
> +  for (int i=0; i<5; i++) {

  Yes this is a bug :(  Because you are at it, please check the same 
thing at TkStatus.cc file in the same package. Same problem might be 
there (As I recall I copied it from there).

thanks
sushil
> +    if (!dcsTk[i]) {
>
> @sushilchauhan FYI. You seem to be the author of this code so maybe you would be qualified to answer this question.
>

Dr15Jones committed:
Fix bug in status check

Dr15Jones committed:
Make TkStatus an edm::one::EDAnalyzer

-changed to edm::one::EDAnalyzer
-removed unused member variables and functions
-used modern C++ constructs

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
@sushilchauhan I fixed the bug on both modules and updated TkStatus.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24336/6102

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #24336 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

cmsbuild commented:
+1
Tested at: 4115f80778ef270f1df8bda432b9b44c48854742
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24336/29936/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24336/29936/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2983602
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2983410
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

schneiml commented:
@sushilchauhan do you still maintain this module/the online `beam_dqm` client? In that case, we might ask you to validate these changes at some point.

@dinardo your name is also in the history of this file and I know you maintain a beamspot-related online DQM client. So I figured, you might have an opion here...

@Dr15Jones keep in mind that this is online-only, so the PR tests will not do anything useful, and validation is hard (though not impossible) on a non-production release. While I trust you on technical changes, we should ask the people who use this module whenever there is a risk of functional changes. Not sure what sort of procedure we should follow here -- make a backport and test in playback (maybe without ever merging the backport), remember to test once 10_3 hits playback (which will be for HI, as I understand, but we'll probably still playback some pp data in the beginning), or do purely local validation (there seems to be a test script that looks like it might do the right thing).

sushilchauhan commented:
On 21/08/18 2:18 PM, Marcel Schneider wrote:
> @sushilchauhan do you still maintain this module/the online `beam_dqm` client? In that case, we might ask you to validate these changes at some point.
  No.... gave up this task  two years ago.

  sushil
>
> @dinardo your name is also in the history of this file and I know you maintain a beamspot-related online DQM client. So I figured, you might have an opion here...
>
> @Dr15Jones keep in mind that this is online-only, so the PR tests will not do anything useful, and validation is hard (though not impossible) on a non-production release. While I trust you on technical changes, we should ask the people who use this module whenever there is a risk of functional changes. Not sure what sort of procedure we should follow here -- make a backport and test in playback (maybe without ever merging the backport), remember to test once 10_3 hits playback (which will be for HI, as I understand, but we'll probably still playback some pp data in the beginning), or do purely local validation (there seems to be a test script that looks like it might do the right thing).
>

andrius-k commented:
Hi Mauro,

Yes, as Marcel mentioned you appear to be in git history of some of the changes in this PR, so I thought you might have something to comment here.

Dr15Jones commented:
I used the test described here for this module as well:
https://github.com/cms-sw/cmssw/pull/24346#issuecomment-414791372

The sets of plots with and without this change are identical. The temporary files the DQSFileSaverOnline creates each LuminosityBlock also contain the same histograms in both cases.

andrius-k commented:
+1
Ferenc from the BeamSpot team has no concerns about these changes.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

kpedro88 commented:
+1
